Data_Structures_Answers.md

```diff
@@ -5,14 +5,19 @@ Add your answers to the questions below.
 2. What is the space complexity of your `depth_first_for_each` function?
 
 3. What is the runtime complexity of your `breadth_first_for_each` method?
+    I'd say it is O(n). It is visiting every node only once. 
 
 4. What is the space complexity of your `breadth_first_for_each` method?
-
+    O(n). The size of the queue is scaling with the size of the tree. 
 
 5. What is the runtime complexity of the provided code in `names.py`?
+    O(n**2) due to the nested loops. It to just over 10 seconds to run!
 
 6. What is the space complexity of the provided code in `names.py`?
+    O(n). The array of duplicates will increase with the size of the files. 
 
 7. What is the runtime complexity of your optimized code in `names.py`?
+    O(n) Linear. Is only looping once and now only takes 1.5 seconds to run on my computer.
```

Comment:
the current solution is still n**2 though. Python's functions underneath must be doing something much more efficient, but you're still looping through each list twice. 
the check for `if name_1 in some_list` still has to go through every item in the list.